File: Search/RabinKarpAlgorithm.cpp
/*
 * C++ Program to Implement Rabin-Karp Algorithm
 */
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cstdlib>
using namespace std;
#define d 256

void search(char *pat, char *txt, int q)
{
    int M = strlen(pat);
    int N = strlen(txt);
    int i, j;
    int p = 0;
    int t = 0;
    int h = 1;
    for (i = 0; i < M - 1; i++)
        h = (h * d) % q;
    for (i = 0; i < M; i++)
    {
        p = (d *p + pat[i]) % q;
        t = (d * t + txt[i]) % q;
    }
    for (i = 0; i <= N - M; i++)
    {
        if (p == t)
        {
            for (j = 0; j < M; j++)
            {
                if (txt[i + j] != pat[j])
                    break;
            }
            if (j == M)
            {
                cout<<"Pattern found at index: "<<i<<endl;
            }
        }
        if (i < N - M)
        {
            t = (d * (t - txt[i] * h) + txt[i + M]) % q;
            if (t < 0)
              t = (t + q);
        }
    }
}
 
int main()
{
    char *txt = "Rabin Karp Algorithm";
    char *pat = "arp";
    int q = 101;
    search(pat, txt, q);
    return 0;
}

File: String/Rabin-Karp Algorithm.js

// Following program is a Javascript implementation 
// of Rabin Karp Algorithm given in the CLRS book 

// d is the number of characters in
// the input alphabet 
let d = 256; 

/* pat -> pattern 
	txt -> text 
	q -> A prime number 
*/
function search(pat, txt, q) 
{ 
	let M = pat.length; 
	let N = txt.length; 
	let i, j; 
	
	// Hash value for pattern 
	let p = 0; 
	
	// Hash value for txt 
	let t = 0; 
	let h = 1; 

	// The value of h would be "pow(d, M-1)%q" 
	for(i = 0; i < M - 1; i++) 
		h = (h * d) % q; 

	// Calculate the hash value of pattern and 
	// first window of text 
	for(i = 0; i < M; i++) 
	{ 
		p = (d * p + pat[i].charCodeAt()) % q; 
		t = (d * t + txt[i].charCodeAt()) % q; 
	} 

	// Slide the pattern over text one by one 
	for(i = 0; i <= N - M; i++) 
	{ 

		// Check the hash values of current
		// window of text and pattern. If the
		// hash values match then only 
		// check for characters one by one 
		if (p == t) 
		{ 
			
			/* Check for characters one by one */
			for(j = 0; j < M; j++) 
			{ 
				if (txt[i+j] != pat[j]) 
					break; 
			} 

			// if p == t and pat[0...M-1] = 
			// txt[i, i+1, ...i+M-1] 
			if (j == M) 
				document.write("Pattern found at index " + 
								i + "<br/>"); 
		} 

		// Calculate hash value for next window 
		// of text: Remove leading digit, add 
		// trailing digit 
		if (i < N - M) 
		{ 
			t = (d * (t - txt[i].charCodeAt() * h) + 
					txt[i + M].charCodeAt()) % q; 

			// We might get negative value of t, 
			// converting it to positive 
			if (t < 0) 
				t = (t + q); 
		} 
	} 
} 

// Driver code
let txt = "GEEKS FOR GEEKS";
let pat = "GEEK";

// A prime number
let q = 101; 

// Function Call
search(pat, txt, q);


